File: src/main/java/org/cimsbioko/syncadpt/AuthUtils.java
package org.cimsbioko.syncadpt;

import android.content.Context;
import android.os.Build;
import org.json.JSONException;
import org.json.JSONObject;
import org.cimsbioko.R;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

import static android.os.Build.*;
import static android.os.Build.VERSION.RELEASE;
import static java.util.Arrays.asList;
import static org.cimsbioko.utilities.HttpUtils.encodeBasicCreds;
import static org.cimsbioko.utilities.IOUtils.close;
import static org.cimsbioko.utilities.IOUtils.copy;
import static org.cimsbioko.utilities.UrlUtils.buildServerUrl;

public class AuthUtils {

    /**
     * Calls server web api to register the device (initial association).
     *
     * @param ctx      used to get the registration endpoint relative to server url in settings
     * @param name     the device 'username' to use when authenticating to device api
     * @param secret the secret to use when authenticating
     * @return a {@link JSONObject} containing the response data.
     * @throws IOException   when url is bad, or io fails
     * @throws JSONException when construction of the response object fails
     */
    public static JSONObject register(Context ctx, String name, String secret) throws IOException, JSONException {
        HttpURLConnection urlConn = (HttpURLConnection) getRegistrationEndpoint(ctx).openConnection();
        urlConn.setRequestProperty("Content-Type", "application/json");
        urlConn.setRequestProperty("Authorization", encodeBasicCreds(name, secret));
        BufferedOutputStream out = null;
        BufferedInputStream in = null;
        try {
            urlConn.setDoOutput(true);
            urlConn.setChunkedStreamingMode(0);
            out = new BufferedOutputStream(urlConn.getOutputStream());
            out.write(getDeviceDescription().getBytes());
            out.flush();
            if (urlConn.getResponseCode() != HttpURLConnection.HTTP_OK || !urlConn.getContentType().startsWith("application/json")) {
                throw new RuntimeException(String.format(
                        "unexpected response: status %s, mime type = %s",
                        urlConn.getResponseCode(), urlConn.getContentType()));
            } else {
                ByteArrayOutputStream bout = new ByteArrayOutputStream();
                in = new BufferedInputStream(urlConn.getInputStream());
                copy(in, bout);
                return new JSONObject(bout.toString());
            }
        } finally {
            close(out, in);
            urlConn.disconnect();
        }
    }

    /**
     * Calls server web api to obtain a new access token for the device, in other words - a refreshed token.
     *
     * @param ctx      used to get the token endpoint relative to server url in settings
     * @param name     the device 'username' to use when authenticating to device api
     * @param secret the secret to use when authenticating
     * @return a {@link JSONObject} containing the response data.
     * @throws IOException   when url is bad, or io fails
     * @throws JSONException when construction of the response object fails
     */
    public static JSONObject token(Context ctx, String name, String secret) throws IOException, JSONException {
        HttpURLConnection urlConn = (HttpURLConnection) getTokenEndpoint(ctx).openConnection();
        urlConn.setRequestMethod("POST");
        urlConn.setRequestProperty("Content-Type", "application/json");
        urlConn.setRequestProperty("Authorization", encodeBasicCreds(name, secret));
        BufferedInputStream in = null;
        try {
            if (urlConn.getResponseCode() != HttpURLConnection.HTTP_OK || !urlConn.getContentType().startsWith("application/json")) {
                throw new RuntimeException(String.format(
                        "unexpected response: status %s, mime type = %s",
                        urlConn.getResponseCode(), urlConn.getContentType()));
            } else {
                ByteArrayOutputStream bout = new ByteArrayOutputStream();
                in = new BufferedInputStream(urlConn.getInputStream());
                copy(in, bout);
                return new JSONObject(bout.toString());
            }
        } finally {
            close(in);
            urlConn.disconnect();
        }
    }

    /**
     * Returns a device description suitable for general differentiation of device types. This is used to help device
     * administrators identify and manage registered devices with more information than just the device name.
     *
     * @return a description including manufacturer, brand, model, and Android release.
     * @throws JSONException
     */
    public static String getDeviceDescription() throws JSONException {
        JSONObject json = new JSONObject();
        StringBuilder b = new StringBuilder();
        for (String s : asList(MANUFACTURER, BRAND, Build.MODEL, String.format("Android %s", RELEASE))) {
            if (!UNKNOWN.equals(s)) {
                b.append(s);
                b.append("\n");
            }
        }
        return json.put("description", b.toString()).toString();
    }

    /**
     * Returns the server {@link URL} to use to register as a device. It is constructed based on the application's
     * configured server endpoint.
     *
     * @param ctx application context to use for relevant config values
     * @return a {@link URL} object corresponding to the device registration endpoint
     * @throws MalformedURLException when the constructed value is not a valid URL
     */
    public static URL getRegistrationEndpoint(Context ctx) throws MalformedURLException {
        return new URL(buildServerUrl(ctx, ctx.getString(R.string.auth_registration_path)));
    }

    /**
     * Returns the server {@link URL} to use to issue access tokens as a device. It is constructed relative to the
     * application's configured server endpoint.
     *
     * @param ctx application context to use for relevant config values
     * @return a {@link URL} object corresponding to the device token endpoint
     * @throws MalformedURLException when the constructed value is not a valid URL
     */
    public static URL getTokenEndpoint(Context ctx) throws MalformedURLException {
        return new URL(buildServerUrl(ctx, ctx.getString(R.string.auth_token_path)));
    }
}

File: src/main/java/org/cimsbioko/syncadpt/AuthUtils.kt
package org.cimsbioko.syncadpt

import android.content.Context
import android.os.Build
import android.os.Build.VERSION
import org.cimsbioko.R
import org.cimsbioko.utilities.HttpUtils.encodeBasicCreds
import org.cimsbioko.utilities.IOUtils.copy
import org.cimsbioko.utilities.UrlUtils.buildServerUrl
import org.cimsbioko.utilities.use
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

object AuthUtils {

    /**
     * Calls server web api to register the device (initial association).
     *
     * @param ctx      used to get the registration endpoint relative to server url in settings
     * @param name     the device 'username' to use when authenticating to device api
     * @param secret the secret to use when authenticating
     * @return a [JSONObject] containing the response data.
     * @throws IOException   when url is bad, or io fails
     * @throws JSONException when construction of the response object fails
     */
    @JvmStatic
    @Throws(IOException::class, JSONException::class)
    fun register(ctx: Context, name: String, secret: String) = getRegistrationEndpoint(ctx).openConnection()
            .apply {
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", encodeBasicCreds(name, secret))
            }.let { c ->
                with(c as HttpURLConnection) {
                    try {
                        doOutput = true
                        setChunkedStreamingMode(0)
                        BufferedOutputStream(outputStream).use {
                            it.apply {
                                write(deviceDescription.toByteArray())
                                flush()
                            }
                        }
                        toJsonObject()
                    } finally {
                        disconnect()
                    }
                }
            }


    /**
     * Calls server web api to obtain a new access token for the device, in other words - a refreshed token.
     *
     * @param ctx      used to get the token endpoint relative to server url in settings
     * @param name     the device 'username' to use when authenticating to device api
     * @param secret the secret to use when authenticating
     * @return a [JSONObject] containing the response data.
     * @throws IOException   when url is bad, or io fails
     * @throws JSONException when construction of the response object fails
     */
    @JvmStatic
    @Throws(IOException::class, JSONException::class)
    fun token(ctx: Context, name: String, secret: String) = getTokenEndpoint(ctx).openConnection()
            .let { it as HttpURLConnection }
            .apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", encodeBasicCreds(name, secret))
            }
            .let {
                try {
                    it.toJsonObject()
                } finally {
                    it.disconnect()
                }
            }

    /**
     * Returns a device description suitable for general differentiation of device types. This is used to help device
     * administrators identify and manage registered devices with more information than just the device name.
     *
     * @return a description including manufacturer, brand, model, and Android release.
     * @throws JSONException
     */
    @get:Throws(JSONException::class)
    val deviceDescription: String
        get() = JSONObject()
                .put("description", buildString {
                    for (s in listOf(Build.MANUFACTURER, Build.BRAND, Build.MODEL, "Android ${VERSION.RELEASE}")) {
                        if (Build.UNKNOWN != s) {
                            append(s)
                            append("\n")
                        }
                    }
                })
                .toString()

    /**
     * Returns the server [URL] to use to register as a device. It is constructed based on the application's
     * configured server endpoint.
     *
     * @param ctx application context to use for relevant config values
     * @return a [URL] object corresponding to the device registration endpoint
     * @throws MalformedURLException when the constructed value is not a valid URL
     */
    @Throws(MalformedURLException::class)
    fun getRegistrationEndpoint(ctx: Context): URL = URL(buildServerUrl(ctx, ctx.getString(R.string.auth_registration_path)))

    /**
     * Returns the server [URL] to use to issue access tokens as a device. It is constructed relative to the
     * application's configured server endpoint.
     *
     * @param ctx application context to use for relevant config values
     * @return a [URL] object corresponding to the device token endpoint
     * @throws MalformedURLException when the constructed value is not a valid URL
     */
    @Throws(MalformedURLException::class)
    fun getTokenEndpoint(ctx: Context): URL = URL(buildServerUrl(ctx, ctx.getString(R.string.auth_token_path)))
}

private fun HttpURLConnection.toJsonObject(): JSONObject =
        if (responseCode != HttpURLConnection.HTTP_OK || !contentType.startsWith("application/json")) {
            throw RuntimeException("unexpected response: status $responseCode, mime type = $contentType")
        } else {
            val bis = BufferedInputStream(inputStream)
            val bout = ByteArrayOutputStream()
            listOf(bis, bout).use {
                copy(bis, bout)
                JSONObject(bout.toString())
            }
        }

